cudax/include/cuda/experimental/__stf/internal/parallel_for_scope.cuh

```diff
@@ -60,59 +63,367 @@ __global__ void loop(const _CCCL_GRID_CONSTANT size_t n, shape_t shape, F f, tup
 
   // Help the compiler which may not detect that a device lambda is calling a device lambda
   CUDASTF_NO_DEVICE_STACK
-  auto explode_args = [&](auto... data) {
+  auto explode_args = [&](auto&&... data) {
+    CUDASTF_NO_DEVICE_STACK
+    auto const explode_coords = [&](auto&&... coords) {
+      f(coords..., data...);
+    };
     // For every linearized index in the shape
     for (; i < n; i += step)
     {
-      CUDASTF_NO_DEVICE_STACK
-      auto explode_coords = [&](auto... coords) {
-        f(coords..., data...);
-      };
       ::std::apply(explode_coords, shape.index_to_coords(i));
     }
   };
   ::std::apply(explode_args, mv(targs));
 }
 
+// Create SelectType Using Partial Specialization
+template <typename Oi, typename Ai>
+struct get_owning_container_of
+{
+  using type = typename owning_container_of<Ai>::type; // Default case
+};
+
+template <typename Ai>
+struct get_owning_container_of<::std::monostate, Ai>
+{
+  using type = ::std::monostate;
+};
+
+/**
+ * @brief This wraps tuple of arguments and operators into a class that stores
+ * a tuple of arguments which include local variables for reductions.
+ *
+ * Providing a dedicated class makes it possible to implement reduction, copy
+ * or init operators directly on top of the tuple, which was tedious otherwise.
+ */
+template <typename tuple_args, typename tuple_ops>
+class redux_vars
+{
+  /**
+   * @brief Tuple of arguments needed to store temporary variables used in reduction operations.
+   *
+   * For example, if we have ArgsTuple=tuple<slice<T>, slice<T>, scalar<T>, scalar<U>> and OpsTuple=tuple<none, none,
+   * sum<T>, sum<U>> we will have a type that is tuple<::std::monostate, ::std::monostate, T, U> which corresponds to
+   * the variables we need to store to perform reductions.
+   */
+  template <typename ArgsTuple, typename OpsTuple>
+  struct redux_buffer_tup;
+
+  template <typename... Ai, typename... Oi>
+  struct redux_buffer_tup<::std::tuple<Ai...>, ::std::tuple<Oi...>>
+  {
+    static_assert(sizeof...(Ai) == sizeof...(Oi), "Tuples must be of the same size");
+
+    using type = ::cuda::std::tuple<typename get_owning_container_of<typename Oi::first_type, Ai>::type...>;
+  };
```

Comment:
Question:

This is mixing std::tuple and cuda::std::tuple, is this intentional?

Reply:
Yes, it's unfortunate but intended : we have started to move from std::tuple to cuda::std::tuple in the generated code, but most of our types are still std::tuple, so that's where we "start" to transition.